pool.go

```diff
@@ -0,0 +1,336 @@
+package server
+
+import (
+	"fmt"
+	"math"
+	"sync/atomic"
+	"time"
+
+	"github.com/bblfsh/sdk/protocol"
+	"github.com/pkg/errors"
+)
+
+var (
+	// DefaultScalingPolicy is the default ScalingPolicy.
+	DefaultScalingPolicy = MovingAverage(10, MinMax(1, 10, AIMD(1, 0.5)))
+	// DefaultPoolTimeout is the time a request to the DriverPool can wait
+	// before getting a driver assigned.
+	DefaultPoolTimeout = time.Second * 5
+)
+
+// DriverPool controls a pool of drivers and balances requests among them,
+// ensuring each driver does not get concurrent requests. The number of driver
+// instances in the driver pool is controlled by a ScalingPolicy.
+type DriverPool struct {
+	scalingPolicy ScalingPolicy
+	timeout       time.Duration
+	new           func() (Driver, error)
+
+	cur        int
+	close      chan struct{}
+	closed     bool
+	waiting    *atomicInt
+	readyQueue *driverQueue
+}
+
+// StartDriverPool creates and starts a new DriverPool. It takes as parameters
+// the ScalingPolicy, the timeout for requests before getting a driver assigned,
+// and a function to create driver instances.
+func StartDriverPool(scaling ScalingPolicy, timeout time.Duration, new func() (Driver, error)) (*DriverPool, error) {
+	dp := &DriverPool{
+		scalingPolicy: scaling,
+		timeout:       timeout,
+		new:           new,
+		close:         make(chan struct{}),
+		waiting:       &atomicInt{},
+		readyQueue:    newDriverQueue(1000),
+	}
+
+	if err := dp.start(); err != nil {
+		_ = dp.Close()
+		return nil, err
+	}
+
+	return dp, nil
+}
+
+func (dp *DriverPool) start() error {
+	target := dp.scalingPolicy.Scale(0, 0)
+	if err := dp.setInstanceCount(target); err != nil {
+		return err
+	}
+
+	go dp.scaling()
+	return nil
+}
+
+// setInstanceCount changes the number of running driver instances. Instances
+// will be started or stopped as necessary to satisfy the new instance count.
+// It blocks until the all required instances are started or stopped.
+func (dp *DriverPool) setInstanceCount(target int) error {
+	if target < 0 {
+		return fmt.Errorf("cannot set instances to negative number")
+	}
+
+	n := target - dp.cur
+	if n > 0 {
+		return dp.add(n)
+	} else if n < 0 {
+		return dp.del(-n)
+	}
+
+	return nil
+}
+
+func (dp *DriverPool) add(n int) error {
+	for i := 0; i < n; i++ {
+		d, err := dp.new()
+		if err != nil {
+			return err
+		}
+
+		dp.readyQueue.Enqueue(d)
+		dp.cur++
+	}
+
+	return nil
+}
+
+func (dp *DriverPool) del(n int) error {
+	for i := 0; i < n; i++ {
+		d, more := dp.readyQueue.Dequeue()
+		if !more {
+			return fmt.Errorf("driver queue closed")
+		}
+
+		dp.cur--
+		if err := d.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (dp *DriverPool) scaling() {
+	for {
+		time.Sleep(time.Millisecond * 100)
```

Comment:
change this to a select on the close chanel and a time.After so this is responsive instead of sleeping.

You will need to change the close channel semantics, but I think that will also be an improvement, in the `Close` method you can send `dp.close <- "please scaling gorutine, die now"` and on this function, the select should react with sending `dp.close <- "scaling is dead for good, you can proceed with the closing"`.